drivers/modem/gsm_ppp.c

```diff
@@ -669,6 +719,49 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
 		return;
 	}
 
+registering:
+	/* Wait for cell tower registration */
+	ret = modem_cmd_send_nolock(&gsm->context.iface,
+				    &gsm->context.cmd_handler,
+				    &check_net_reg_cmd, 1,
+				    "AT+CREG?",
+				    &gsm->sem_response,
+				    GSM_CMD_SETUP_TIMEOUT);
+	if ((ret < 0) || ((gsm->net_state != GSM_ROAMING) &&
+			 (gsm->net_state != GSM_HOME_NETWORK))) {
+		if (!gsm->register_retries) {
+			gsm->register_retries = CONFIG_MODEM_GSM_REGISTER_TIMEOUT *
+				MSEC_PER_SEC / GSM_REGISTER_DELAY_MSEC;
+		} else {
+			gsm->register_retries--;
+
+			/* Reset RF if timed out */
+			if (!gsm->register_retries) {
+				(void)modem_cmd_send_nolock(&gsm->context.iface,
+							    &gsm->context.cmd_handler,
+							    &response_cmds[0],
+							    ARRAY_SIZE(response_cmds),
+							    "AT+CFUN=0", &gsm->sem_response,
+							    GSM_CMD_AT_TIMEOUT);
+
+				k_msleep(GSM_REGISTER_DELAY_MSEC);
+
+				(void)modem_cmd_send_nolock(&gsm->context.iface,
```

Comment:
in fact I'm not 100% convinced this is the correct action if registration times out. What if the modem isn't responding to AT commands at all? We would want to power cycle the modem in that case.

To me, it would seem more natural to fail and signal this to the app somehow. But I'm not sure how, since I don't think there's a net_event for that. Perhaps it's time we did a blocking version of `gsm_ppp_start()`?

Reply:
> shouldn't you reset net_state here somewhere if it fails?

No need, it is set by the handler of the AT+CREG

Reply:
> Perhaps it's time we did a blocking version of `gsm_ppp_start()`?

Not sure about this, what I do in my code is to wait for the net connection event with a timeout, and restart the driver when timed out

Reply:
You are right, but it is reset in `gsm_ppp_stop`, and unsolicited output is disabled for this: https://github.com/zephyrproject-rtos/zephyr/blob/4a060c938029475140fddbe1d560dc7e08e69f33/drivers/modem/gsm_ppp.c#L476

I guess nothing except the `AT+CREG?` can set it to something else other than its reset value?

Reply:
> > Perhaps it's time we did a blocking version of `gsm_ppp_start()`?
> 
> Not sure about this, what I do in my code is to wait for the net connection event with a timeout, and restart the driver when timed out

Ofc I could be wrong, but it doesn't look like you're clearing all state on all possible failures. What if net_state == SEARCHING, then we time out and reach max retries? Looks like net_state == SEARCHING will still be true

Reply:
> Not sure about this, what I do in my code is to wait for the net connection event with a timeout, and restart the driver when timed out

It seems that we are putting lot of needed connection logic into the application. @ycsin would it be possible if you could add something to gsm_modem sample app in order to make it somewhat usable example?

Reply:
But... you reschedule the work item. Once it gets executed again, won't it see the state and `goto registering`?